### PR TITLE
feat: agent auth

### LINKED ISSUE #60: agent auth
Image

### PR DESCRIPTION
**Pull request by SzymonSt**
Resolves #60

### CODE REVIEW

File: backend/pkg/utils/helpers.go [hunk 1]
```diff
@@ -107,3 +109,13 @@ func FilterForRelevantLogs(logs []string) (relevantLogs []string) {
 
 	return
 }
+
+func GenerateBearerToken() (string, error) {
+	b := make([]byte, 32)
```
Comment (MRadzynski) [inline backend/pkg/utils/helpers.go:114]: please, provide a better name for this variable

File: backend/pkg/utils/helpers.go [hunk 2]
```diff
@@ -107,3 +109,13 @@ func FilterForRelevantLogs(logs []string) (relevantLogs []string) {
...
+	_, err := rand.Read(b)
+	if err != nil {
+		return "", err
+	}
+	token := base64.StdEncoding.EncodeToString(b)
```
Comment (MRadzynski) [inline backend/pkg/utils/helpers.go:119]: plesae, add an empty line here

File: ext/frontend/src/app/auth/services/auth-state.service.ts [hunk 3]
```diff
@@ -257,13 +257,22 @@ export class AuthStateService implements OnDestroy {
     this.router.navigateByUrl('/login')
     this.wasNavigatedFromLogin = false;
   }
+  private async authenthicateAndRunAgent():Promise<void> {
+    const agentToken = await this.authService.authAgent(this.token).toPromise().then((result: { agentToken: string }) => {
+      return result.agentToken;
+    })
+    this.configurationService.setAgentAuthData({token: agentToken, userId: this.userData.id})
```
Comment (MRadzynski) [inline ext/frontend/src/app/auth/services/auth-state.service.ts:264]: please, add an empty line here

File: ext/frontend/src/app/auth/services/auth.service.ts [hunk 4]
```diff
@@ -67,6 +67,21 @@ export class AuthService {
       );
   }
 
+  public authAgent(token: Token): Observable<{ agentToken: string }> {
+    const headers: HttpHeaders = new HttpHeaders({
+      'Content-Type': 'application/json',
```
Comment (MRadzynski) [inline ext/frontend/src/app/auth/services/auth.service.ts:72]: please, alpha-sort the headers to make them easier to check if some header is already added in the future

File: ext/frontend/src/app/shared/interfaces/AgentAuthDataDTO.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export class agentAuthDataDTO {
+    public constructor(
+        public userId: number, 
+        public token: string) {
+    }
+}
```
Comment (MRadzynski) [inline ext/frontend/src/app/shared/interfaces/AgentAuthDataDTO.ts:6]: try to add an empty line at the end of the file

File: ext/frontend/src/app/shared/services/configuration.service.ts [hunk 6]
```diff
@@ -40,4 +41,10 @@ export class ConfigurationService {
 
     });
   }
+
+  public setAgentAuthData(agentAuthData: agentAuthDataDTO): void {
+    this.httpClient.post<void>(`${environment.agentApiUrl}/control/auth_data`, agentAuthData).subscribe(() => {
+      this.toastrService.success(this.translateService.instant('CONFIGURATION.AGENT_AUTH_DATA_UPDATED'));
+    });
+  }
 }
```
Comment (MRadzynski) [inline ext/frontend/src/app/shared/services/configuration.service.ts:50]: try to add an empty line at the end of the file

File: backend/pkg/controllers/main.controller.go [hunk 7]
```diff
@@ -681,6 +680,69 @@ func (c *MainController) RefreshTokenHandler(ctx *gin.Context) {
...
+		return
+	}
+
+	ctx.JSON(http.StatusOK, gin.H{
+		"message": "Success",
+		"token":   token,
+	})
+}
+
+func (c *MainController) CheckAgentAuthorization(ctx *gin.Context) {
```
Comment (MRadzynski) [inline backend/pkg/controllers/main.controller.go:730]: as this is a middleware please move it to the middleware directory

File: backend/pkg/controllers/main.controller.go [hunk 8]
```diff
@@ -681,6 +680,69 @@ func (c *MainController) RefreshTokenHandler(ctx *gin.Context) {
 	})
 }
 
+func (c *MainController) AuthenticateAgent(ctx *gin.Context) {
+	var user models.User
+	userId, err := getUserIdFromToken(ctx)
```
Comment (MRadzynski) [inline backend/pkg/controllers/main.controller.go:685]: please add an empty line to separate the variables declaration from the function body

File: backend/pkg/controllers/main.controller.go [hunk 9]
```diff
@@ -681,6 +680,69 @@ func (c *MainController) RefreshTokenHandler(ctx *gin.Context) {
...
+	if err != nil {
+		ctx.JSON(http.StatusUnauthorized, gin.H{"error": err.Error()})
+		return
+	}
+
+	result := c.usersCollection.FindOne(ctx, bson.M{"userId": userId})
+	err = result.Decode(&user)
+	if err != nil {
+		fmt.Printf("Error: %s", err)
+		ctx.JSON(400, gin.H{"error": err.Error()})
```
Comment (MRadzynski) [inline backend/pkg/controllers/main.controller.go:695]: please change the number code to the constants from `http` package

File: backend/pkg/controllers/main.controller.go [hunk 10]
```diff
@@ -681,6 +680,69 @@ func (c *MainController) RefreshTokenHandler(ctx *gin.Context) {
...
+		return
+	}
+	if user.AgentBearerToken != "" {
```
Comment (MRadzynski) [inline backend/pkg/controllers/main.controller.go:698]: add an empty line to improve readability

File: backend/pkg/controllers/main.controller.go [hunk 11]
```diff
@@ -681,6 +680,69 @@ func (c *MainController) RefreshTokenHandler(ctx *gin.Context) {
...
+		ctx.JSON(http.StatusOK, gin.H{
+			"message": "Success",
+			"token":   user.AgentBearerToken,
```
Comment (MRadzynski) [inline backend/pkg/controllers/main.controller.go:701]: Doesn't the agent need the refresh token?

**Review by MRadzynski — CHANGES_REQUESTED** [review]
Awesome! 🤩  Just minor adjustments

File: backend/pkg/controllers/main.controller.go (hunk 7 above)
Comment (SzymonSt) [reply backend/pkg/controllers/main.controller.go:730]: Cannot do, I need access to main controller to query database, I don't know any workaround to that

File: backend/pkg/controllers/main.controller.go (hunk 11 above)
Comment (SzymonSt) [reply backend/pkg/controllers/main.controller.go:701]: nope agent token exp is not verified it technically never expires

File: ext/frontend/src/app/auth/services/auth.service.ts [hunk 12]
```diff
@@ -67,6 +67,21 @@ export class AuthService {
       );
   }
 
+  public authAgent(token: Token): Observable<{ agentToken: string }> {
+    const headers: HttpHeaders = new HttpHeaders({
```
Comment (Paweleksltzn) [inline ext/frontend/src/app/auth/services/auth.service.ts:71]: Tutaj te nagłowki potrzebne są? Wydaje mi sie że interceptor od autha by to ogarnął chyba. Mozna sprobowac to wywalic i zerknac czy przechodzi a jak nie no to zostawic

Comment (SzymonSt) [reply ext/frontend/src/app/auth/services/auth.service.ts:71]: yep it works

**Pull request merged**